File: android/src/main/java/com/mobiletracker/StorageManager.kt
package com.mobiletracker

import android.content.Context
import android.content.SharedPreferences
import java.io.File

/**
 * StorageManager handles persistent storage with dual storage strategy
 * Primary: SharedPreferences (similar to web cookies)
 * Secondary: File backup (similar to web localStorage)
 * 
 * Web Reference: api.ts getCookie(), writeCookie(), clearCookie() methods (lines 37-127)
 */
class StorageManager(
    context: Context,
    private val prefix: String
) {
    private val prefs: SharedPreferences = context.getSharedPreferences("MobileTracker", Context.MODE_PRIVATE)
    private val backupDirectory: File = File(context.filesDir, "tracker_backup")
    
    init {
        // Ensure backup directory exists
        if (!backupDirectory.exists()) {
            backupDirectory.mkdirs()
        }
    }
    
    /**
     * Save a value to both primary (SharedPreferences) and secondary (file backup) storage
     * 
     * @param key Storage key (will be prefixed)
     * @param value String value to store
     * @param expires Optional expiration in days (not enforced, for compatibility)
     */
    fun save(key: String, value: String, expires: Int? = null) {
        val fullKey = prefix + key
        
        // Save to SharedPreferences (primary storage)
        prefs.edit().putString(fullKey, value).apply()
        
        // Save to file backup (secondary storage)
        saveToFileBackup(fullKey, value)
    }
    
    /**
     * Retrieve a value from storage, checking SharedPreferences first, then file backup
     * 
     * @param key Storage key (will be prefixed)
     * @return Stored value if found, null otherwise
     */
    fun retrieve(key: String): String? {
        val fullKey = prefix + key
        
        // Try SharedPreferences first (primary storage)
        prefs.getString(fullKey, null)?.let { return it }
        
        // Fallback to file backup (secondary storage)
        return retrieveFromFileBackup(fullKey)
    }
    
    /**
     * Remove a value from both storages
     * 
     * @param key Storage key (will be prefixed)
     */
    fun remove(key: String) {
        val fullKey = prefix + key
        
        // Remove from SharedPreferences
        prefs.edit().remove(fullKey).apply()
        
        // Remove from file backup
        removeFromFileBackup(fullKey)
    }
    
    /**
     * Clear all keys with the current prefix from both storages
     */
    fun clear() {
        // Clear all keys with prefix from SharedPreferences
        val editor = prefs.edit()
        prefs.all.keys.filter { it.startsWith(prefix) }.forEach { editor.remove(it) }
        editor.apply()
        
        // Clear file backup
        clearFileBackup()
    }
    
    // MARK: - File Backup Methods (Secondary Storage)
    
    /**
     * Save value to file backup
     */
    private fun saveToFileBackup(key: String, value: String) {
        try {
            val file = File(backupDirectory, sanitizeFilename(key))
            file.writeText(value)
        } catch (e: Exception) {
            // Silently fail - backup is optional
            // Could log in debug mode if needed
        }
    }
    
    /**
     * Retrieve value from file backup
     */
    private fun retrieveFromFileBackup(key: String): String? {
        return try {
            val file = File(backupDirectory, sanitizeFilename(key))
            if (file.exists()) {
                file.readText()
            } else {
                null
            }
        } catch (e: Exception) {
            // File doesn't exist or can't be read
            null
        }
    }
    
    /**
     * Remove value from file backup
     */
    private fun removeFromFileBackup(key: String) {
        try {
            val file = File(backupDirectory, sanitizeFilename(key))
            if (file.exists()) {
                file.delete()
            }
        } catch (e: Exception) {
            // Silently fail
        }
    }
    
    /**
     * Clear all files in backup directory with the current prefix
     */
    private fun clearFileBackup() {
        try {
            val sanitizedPrefix = sanitizeFilename(prefix)
            backupDirectory.listFiles()?.forEach { file ->
                if (file.name.startsWith(sanitizedPrefix)) {
                    file.delete()
                }
            }
        } catch (e: Exception) {
            // Silently fail
        }
    }
    
    /**
     * Sanitize key for use as filename
     */
    private fun sanitizeFilename(key: String): String {
        // Replace characters that are invalid in filenames
        return key.replace("/", "_")
            .replace(":", "_")
            .replace("\\", "_")
    }
}

File: ios/MobileTracker/StorageManager.swift
import Foundation

/// StorageManager handles persistent storage with dual storage strategy
/// Primary: UserDefaults (similar to web cookies)
/// Secondary: File backup (similar to web localStorage)
class StorageManager {
    private let prefix: String
    private let userDefaults = UserDefaults.standard
    private let fileManager = FileManager.default
    private let backupDirectory: URL
    
    /// Initialize StorageManager with a storage prefix
    /// - Parameter prefix: Storage key prefix in format "__GT_{brandId}_"
    init(prefix: String) {
        self.prefix = prefix
        
        // Create backup directory in app's documents directory
        let documentsPath = fileManager.urls(for: .documentDirectory, in: .userDomainMask)[0]
        self.backupDirectory = documentsPath.appendingPathComponent("tracker_backup", isDirectory: true)
        
        // Ensure backup directory exists
        try? fileManager.createDirectory(at: backupDirectory, withIntermediateDirectories: true, attributes: nil)
    }
    
    /// Save a value to both primary (UserDefaults) and secondary (file backup) storage
    /// - Parameters:
    ///   - key: Storage key (will be prefixed)
    ///   - value: String value to store
    ///   - expires: Optional expiration in days (not enforced, for compatibility)
    func save(key: String, value: String, expires: Int? = nil) {
        let fullKey = prefix + key
        
        // Save to UserDefaults (primary storage)
        userDefaults.set(value, forKey: fullKey)
        userDefaults.synchronize()
        
        // Save to file backup (secondary storage)
        saveToFileBackup(key: fullKey, value: value)
    }
    
    /// Retrieve a value from storage, checking UserDefaults first, then file backup
    /// - Parameter key: Storage key (will be prefixed)
    /// - Returns: Stored value if found, nil otherwise
    func retrieve(key: String) -> String? {
        let fullKey = prefix + key
        
        // Try UserDefaults first (primary storage)
        if let value = userDefaults.string(forKey: fullKey) {
            return value
        }
        
        // Fallback to file backup (secondary storage)
        return retrieveFromFileBackup(key: fullKey)
    }
    
    /// Remove a value from both storages
    /// - Parameter key: Storage key (will be prefixed)
    func remove(key: String) {
        let fullKey = prefix + key
        
        // Remove from UserDefaults
        userDefaults.removeObject(forKey: fullKey)
        userDefaults.synchronize()
        
        // Remove from file backup
        removeFromFileBackup(key: fullKey)
    }
    
    /// Clear all keys with the current prefix from both storages
    func clear() {
        // Clear all keys with prefix from UserDefaults
        let keys = userDefaults.dictionaryRepresentation().keys
        for key in keys where key.hasPrefix(prefix) {
            userDefaults.removeObject(forKey: key)
        }
        userDefaults.synchronize()
        
        // Clear file backup
        clearFileBackup()
    }
    
    // MARK: - File Backup Methods (Secondary Storage)
    
    /// Save value to file backup
    private func saveToFileBackup(key: String, value: String) {
        let fileURL = backupDirectory.appendingPathComponent(sanitizeFilename(key))
        
        do {
            try value.write(to: fileURL, atomically: true, encoding: .utf8)
        } catch {
            // Silently fail - backup is optional
            if userDefaults.bool(forKey: "debug") {
                print("StorageManager: Failed to save to file backup: \(error)")
            }
        }
    }
    
    /// Retrieve value from file backup
    private func retrieveFromFileBackup(key: String) -> String? {
        let fileURL = backupDirectory.appendingPathComponent(sanitizeFilename(key))
        
        do {
            return try String(contentsOf: fileURL, encoding: .utf8)
        } catch {
            // File doesn't exist or can't be read
            return nil
        }
    }
    
    /// Remove value from file backup
    private func removeFromFileBackup(key: String) {
        let fileURL = backupDirectory.appendingPathComponent(sanitizeFilename(key))
        
        try? fileManager.removeItem(at: fileURL)
    }
    
    /// Clear all files in backup directory with the current prefix
    private func clearFileBackup() {
        guard let files = try? fileManager.contentsOfDirectory(at: backupDirectory, includingPropertiesForKeys: nil) else {
            return
        }
        
        let sanitizedPrefix = sanitizeFilename(prefix)
        for fileURL in files {
            let filename = fileURL.lastPathComponent
            if filename.hasPrefix(sanitizedPrefix) {
                try? fileManager.removeItem(at: fileURL)
            }
        }
    }
    
    /// Sanitize key for use as filename
    private func sanitizeFilename(_ key: String) -> String {
        // Replace characters that are invalid in filenames
        return key.replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: ":", with: "_")
            .replacingOccurrences(of: "\\", with: "_")
    }
}
